### PR TITLE
[Spark-9769][ML][PY] add python api for countvectorizermodel

### PR DESCRIPTION
**Pull request by holdenk**
From JIRA: Add Python API, user guide and example for ml.feature.CountVectorizerModel

### CODE REVIEW

File: python/pyspark/ml/feature.py [hunk 1]
```diff
@@ -167,6 +168,133 @@ def getSplits(self):
...
+    minDF = Param(
+        Params._dummy(), "minDF", "Specifies the minimum number of" +
+        " different documents a term must appear in to be included in the vocabulary." +
+        " If this is an integer >= 1, this specifies the number of documents the term must" +
+        " appear in; if this is a double in [0,1), then this specifies the fraction of documents.")
+    vocabSize = Param(
+        Params._dummy(), "vocabSize", "max size of the vocabulary")
+
+    @keyword_only
+    def __init__(self, minTF=None, minDF=None, vocabSize=None, inputCol=None, outputCol=None):
```
Comment (mengxr) [inline python/pyspark/ml/feature.py]: please add default values for `minTF`, `minDF` and `vocabSize`

File: python/pyspark/ml/feature.py [hunk 2]
```diff
@@ -167,6 +168,133 @@ def getSplits(self):
...
+            "fitting.")
+        minDF = Param(
+            self, "minDF", "Specifies the minimum number of" +
+            " different documents a term must appear in to be included in the vocabulary." +
+            " If this is an integer >= 1, this specifies the number of documents the term must" +
+            " appear in; if this is a double in [0,1), then this specifies the fraction of " +
+            "documents.")
+        vocabSize = Param(
+            self, "vocabSize", "max size of the vocabulary")
+        kwargs = self.__init__._input_kwargs
```
Comment (mengxr) [inline python/pyspark/ml/feature.py]: and `_setDefault` here.

**Comment by holdenk** [comment]
jenkins, retest this please.

File: python/pyspark/ml/feature.py [hunk 3]
```diff
@@ -167,6 +168,137 @@ def getSplits(self):
 
 
 @inherit_doc
+class CountVectorizer(JavaEstimator, HasInputCol, HasOutputCol):
+    """
+    Extracts a vocabulary from document collections and generates a [[CountVectorizerModel]].
```
Comment (mengxr) [inline python/pyspark/ml/feature.py]: * Use sphinx link instead of `[[...]]`. You can check the generate HTML doc and see whether the links are broken.
* docstring line width is 72 defined by PEP8.

File: python/pyspark/ml/feature.py [hunk 4]
```diff
@@ -167,6 +168,137 @@ def getSplits(self):
...
+    |0    |[a, b, c]      |(3,[0,1,2],[1.0,1.0,1.0])|
+    |1    |[a, b, b, c, a]|(3,[0,1,2],[2.0,2.0,1.0])|
+    +-----+---------------+-------------------------+
+    ...
+    """
+
+    # a placeholder to make it appear in the generated doc
+    minTF = Param(
+        Params._dummy(), "minTF", "Filter to ignore rare words in" +
+        " a document. For each document, terms with frequency/count less than the given" +
```
Comment (mengxr) [inline python/pyspark/ml/feature.py]: Not part of this PR, but it would be nice to remove this boilerplate code. There should be a way to document the params in `__iniit__()`. I'm thinking of a decorator that can read `param.doc` and put it inside `__doc__`, but I'm not sure whether it would work.

File: python/pyspark/ml/feature.py [hunk 5]
```diff
@@ -167,6 +168,137 @@ def getSplits(self):
...
+        Params._dummy(), "minDF", "Specifies the minimum number of" +
+        " different documents a term must appear in to be included in the vocabulary." +
+        " If this is an integer >= 1, this specifies the number of documents the term must" +
+        " appear in; if this is a double in [0,1), then this specifies the fraction of documents." +
+        " Default 1.0")
+    vocabSize = Param(
+        Params._dummy(), "vocabSize", "max size of the vocabulary. Default 2**18.")
+
+    @keyword_only
+    def __init__(self, minTF=1.0, minDF=1.0, vocabSize=2**18, inputCol=None, outputCol=None):
```
Comment (mengxr) [inline python/pyspark/ml/feature.py]: `2**18` -> `1<<18`, cheaper and consistent with Scala code. Please update the value in other places as well.

File: python/pyspark/ml/feature.py [hunk 6]
```diff
@@ -167,6 +168,137 @@ def getSplits(self):
...
+            "fitting. Default 1.0")
+        self.minDF = Param(
+            self, "minDF", "Specifies the minimum number of" +
+            " different documents a term must appear in to be included in the vocabulary." +
+            " If this is an integer >= 1, this specifies the number of documents the term must" +
+            " appear in; if this is a double in [0,1), then this specifies the fraction of " +
+            "documents. Default 1.0")
+        self.vocabSize = Param(
+            self, "vocabSize", "max size of the vocabulary. Default 2**18.")
+        self._setDefault(minTF=1.0)
```
Comment (mengxr) [inline python/pyspark/ml/feature.py]: `_setDefault` can take multiple arguments.

File: python/pyspark/ml/feature.py [hunk 7]
```diff
@@ -167,6 +168,137 @@ def getSplits(self):
...
+    >>> cv = CountVectorizer(inputCol="raw", outputCol="vectors")
+    >>> model = cv.fit(df)
+    >>> model.transform(df).show(truncate=False)
+    +-----+---------------+-------------------------+
+    |label|raw            |vectors                  |
+    +-----+---------------+-------------------------+
+    |0    |[a, b, c]      |(3,[0,1,2],[1.0,1.0,1.0])|
+    |1    |[a, b, b, c, a]|(3,[0,1,2],[2.0,2.0,1.0])|
+    +-----+---------------+-------------------------+
+    ...
```
Comment (mengxr) [inline python/pyspark/ml/feature.py]: also test `vocabulary`

File: python/pyspark/ml/feature.py (hunk 4 above)
Comment (holdenk) [inline python/pyspark/ml/feature.py]: That sounds like a good improvement for PySpark/ML.

File: python/pyspark/sql/tests.py [hunk 8]
```diff
@@ -830,12 +830,12 @@ def test_infer_long_type(self):
         self.assertEquals(100000000000000, df1.first().f2)
 
         self.assertEqual(_infer_type(1), LongType())
-        self.assertEqual(_infer_type(2**10), LongType())
-        self.assertEqual(_infer_type(2**20), LongType())
-        self.assertEqual(_infer_type(2**31 - 1), LongType())
-        self.assertEqual(_infer_type(2**31), LongType())
-        self.assertEqual(_infer_type(2**61), LongType())
-        self.assertEqual(_infer_type(2**71), LongType())
+        self.assertEqual(_infer_type(1<<10), LongType())
```
Comment (mengxr) [inline python/pyspark/sql/tests.py]: Please remove unrelated content. You can make a new PR for this, though I don't think the change is necessary.

**Comment by mengxr** [comment]
@holdenk Please fix style issues and remove changes to SQL. It looks good to me otherwise.

File: python/pyspark/sql/tests.py (hunk 8 above)
Comment (holdenk) [inline python/pyspark/sql/tests.py]: Ah sorry, I though the earlier comment was to update the 2** to 1<< everywhere in pyspark (since there were not a lot of places.) I'll revert this change.

**Comment by holdenk** [comment]
@mengxr I reverted the SQL changes & the style issue with 1<<

**Comment by mengxr** [comment]
LGTM. Merged into master. Thanks!
